locator/tablets.cc

```diff
@@ -551,7 +551,7 @@ class tablet_effective_replication_map : public effective_replication_map {
         auto& topo = _tmptr->get_topology();
         for (auto&& replica : replicas) {
             auto* node = topo.find_node(replica.host);
-            if (node) {
+            if (node && !node->left()) {
```

Comment:
@tgrabiec shouldn't this check also apply to `to_host_set` below?

Reply:
Hmm, this comment explains why not:
```
    /// Returns replicas for a given token.
    /// During topology change returns replicas which should be targets for writes, excluding the pending replica.
    /// Unlike get_natural_endpoints(), the replica set may include nodes in the left state which were
    /// replaced but not yet rebuilt.
    virtual host_id_vector_replica_set get_replicas(const token& search_token) const = 0;
```

It's quite confusing...  I wish the interface would rely less on those subtle differences and that the function names would reflect those semantics more clearly.
Without reading the comment (on the base class), there is no reason to suspect that get_replicas will return nodes in the left state but get_natural_endpoints won't.

Reply:
There is no reason to suspect that get_replicas() wouldn't return nodes in the left state, since it returns all replicas, some of which may be in the left state.

The reason left nodes are omitted in the ip-based version is that we may not have ips for left nodes. I consider get_natural_endpoints() as legacy awkward API, and will go away after we complete the transition to host ids.

Reply:
removenode waits for tablets to be migrated as part of tablet draining stage of the topology change. This is for node replace, where we rebuild tablet replicas after the node is replaced in the topology. The reason is that we want to support the case of replacing a node in a cluster where RF=N, in which case we cannot drain during topology change because the replacing node is not yet a candidate for replicas. It cannot be due to rollback, which cannot involve migrating tablets back.

Reply:
Maybe we could do replace like this:
- bootstrap the replacing node, stealing the tokens (in the "old token ring") from replaced node -- vnode-based table data is rebuilt on the replacing node
- but keep the replaced node as a "normal" member of the cluster, just without any tokens. So it's still not "left", it just doesn't have any tokens in token ring (but it might be a replica for some tablets)
- now migrate tablets. We're migrating tablets from one normal member to another member, except that the source is marked as permanently dead, so we're rebuilding on the destination
- after migration is finished, only then remove the node moving it to left state


so IIUC the only difference from current procedure is that we prolong the time that a node stays as "normal", and we keep the invariant that "left" nodes are never replicas

I know we don't have "0 token nodes"... yet, actually Patryk is implementing it as part of the "arbiter DC" task